contributors/design-proposals/storage/dynamic_volume_limit.md

```diff
@@ -0,0 +1,67 @@
+# Dynamic attached volume limits
+
+## Goals
+
+Currently the number of volumes attachable to a node is either hardcoded or only configurable via an environment variable. Also
+existing limits only apply to well known volume types like EBS, GCE and is not available to all volume plugins.
+
+This proposal enables any volume plugin to specify those limits and also allows same volume type to have different volume
+limits depending on type of node.
+
+## Implementation Design
+
+### Prerequisite
+
+* This feature will be protected by an alpha feature gate, so as API and CLI changes needed for it. We are planning to call
+  the feature `DynamicVolumeThreshold`.
+
+### API Changes
+
+We propose following changes to node's status field:
+
+```
+node.Status.AttachedVolumeLimits map[string]int{}
+node.Status.AttachedVolumeLimit: {
+  "kubernetes.io/aws-ebs": 39
+}
+```
+
+For CSI the name used in key will be - `kubernetes.io/csi/<driver_name>`
+
+### Changes to CSI
+
+Corresponding change is being made to CSI so as CSI drivers can output
+volume limits as part of Node capabilities - https://github.com/container-storage-interface/spec/pull/221
+
+csi node-registrar will be updated to set the `node.Status` property after fetching
+volume limit from CSI driver.
+
+### Changes to scheduler
+
+When a node is being considered for scheduling of a pod with volume, it will use currently running pods against value of node.Status.AttachedVolumeLimit.
+
+To be backward compatible - the scheduler will fallback to older logic, if `AttachedVolumeLimit` is not set.  This is to make sure that - a updated scheduler can still work with older nodes.
+
+### Setting of limit for existing in-tree volume plugins
+
+The volume limit for existing volume plugins will be set by querying the cloudprovider. Following interface
+will be added to `pkg/cloudprovider/cloud.go`:
+
+```go
+type VolumeLimits interface {
+    GetVolumeLimits(ctx context.Context, nodeName types.NodeName) (map[string]int64, error)
+}
+```
+
+If given cloudprovider implements this interface - kubelet will query the interface and set(via merge) the returned values in
```

Comment:
I believe there's a push towards moving cloud provider dependencies out of kubelet.

Could this be something the external cloud controller calls and updates for each node?